polar-core/src/resource_block.rs

```diff
@@ -298,18 +300,43 @@ impl ResourceBlocks {
         resource: Term,
         declarations: Declarations,
         shorthand_rules: Vec<ShorthandRule>,
-    ) {
-        self.declarations.insert(resource.clone(), declarations);
-        self.shorthand_rules
-            .insert(resource.clone(), shorthand_rules);
+    ) -> PolarResult<()> {
+        // Check for duplicate existing shorthand rules and return a
+        // ValidationError if we discover one
+        if let Some(existing_rules) = self.shorthand_rules.get(&resource) {
+            let existing: HashSet<&ShorthandRule> = existing_rules.iter().collect();
+            for rule in &shorthand_rules {
+                if existing.contains(rule) {
+                    return Err(ValidationError::DuplicateShorthandRule {
+                        resource,
+                        shorthand_rule: rule.clone(),
+                    }
+                    .into());
+                }
+            }
+        }
+
+        // Merge existing declarations if we are reopening a resource block, otherwise add new
+        if let Some(existing) = self.declarations.get_mut(&resource) {
+            existing.extend(declarations.into_iter());
```

Comment:
I don't know this code very well, but when you were presenting was thinking about this - 

What happens if I declare the same term twice?

```
resource Org {
    roles = ["foo"];
}

resource Org {
    permissions = ["foo"];
}
```

I don't think this would be caught with the current code -- seems like this could lead to bugs.

Reply:
good shout. will add a check for this & corresponding unit test